src/styles/text/label.js

```diff
@@ -10,41 +12,81 @@ export default class Label {
             text,
             size,
             position: [],
-            bbox: [],
+            aabb: [],
             move_in_tile,
             keep_in_tile
         });
 
         this.id = Label.id++;
-
         this.buffer = this.buffer || 2; // TODO: make configurable
         this.buffer *= Geo.units_per_pixel;
+        this.keepMinDistance = true;
     }
 
     isComposite () {
         return false;
     }
 
-    occluded (bboxes) {
+    occluded (aabbs) {
         let intersect = false;
-
-        if (bboxes.length > 0) {
-            boxIntersect([this.bbox], bboxes, (i, j) => {
-                intersect = true;
-                return true;
+        let aabb = this.aabb;
+
+        // Broadphase
+        if (aabbs.length > 0) {
+            boxIntersect([this.aabb], aabbs, (i, j) => {
+                // Narrow phase
+                if (OBB.intersect(aabb.obb, aabbs[j].obb)) {
+                    intersect = true;
+                    return true;
+                }
             });
         }
 
+        // No collision on aabb
         if (!intersect) {
-            bboxes.push(this.bbox);
+            aabbs.push(this.aabb);
         }
 
         return intersect;
     }
 
+    // keep a minimal distance between the labels
+    minDistance (aabbs) {
```

Comment:
Maybe make this method name a little more descriptive, I would have assumed it returned a min distance. Like `checkMinDistance()` or something.